util/network-devp2p/src/ip_utils.rs

```diff
@@ -358,6 +362,81 @@ pub fn map_external_address(local: &NodeEndpoint) -> Option<NodeEndpoint> {
 	None
 }
 
+fn search_natpmp(local: &NodeEndpoint) -> Option<NodeEndpoint> {
+	if let SocketAddr::V4(ref local_addr) = local.address {
+		let local_port = local_addr.port();
+		let local_udp_port = local.udp_port;
+
+		let search_gateway_child = ::std::thread::spawn(move || {
+			let get_public_addr = |n: &mut Natpmp| {
```

Comment:
So further down you check that the response is `Response::Gateway`. I think it would be better to move that logic into this closure with proper comments as to why only `Gateway` is the accepted variant.

The same concern applies to the other closures below.

Reply:
done